### PR TITLE
WIP Refactor ComponentManager to create components on demand

### PR DESCRIPTION
**Pull request by kaphacius**
This PR will change the way `ComponentManager` creates components to be on-demand, instead of creating all of the components beforehand, and reusing them later.

### CODE REVIEW

File: AdyenDropIn/Components/PaymentMethodListComponent.swift [hunk 1]
```diff
@@ -46,30 +46,33 @@ internal final class PaymentMethodListComponent: ComponentLoader, PresentableCom
...
+            listItem.imageURL = LogoURLProvider.logoURL(for: paymentMethod, environment: apiContext.environment)
             listItem.trailingText = displayInformation.disclosureText
             listItem.subtitle = displayInformation.subtitle
-            listItem.showsDisclosureIndicator = showsDisclosureIndicator
-            listItem.selectionHandler = { [unowned self, unowned component] in
-                guard !(component is AlreadyPaidPaymentComponent) else { return }
-                self.delegate?.didSelect(component, in: self)
+
+            // Unsure how to handle this
+//            let showsDisclosureIndicator = (component as? PresentableComponent)?.requiresModalPresentation == true
```
Comment (descorp) [inline AdyenDropIn/Components/PaymentMethodListComponent.swift:61]: This shows `>` chevron on list.. we can probably omite this

File: AdyenDropIn/Components/PaymentMethodListComponent.swift [hunk 2]
```diff
@@ -98,9 +101,9 @@ internal final class PaymentMethodListComponent: ComponentLoader, PresentableCom
     /// - Parameter component: The component for which to start a loading animation.
     internal func startLoading(for component: PaymentComponent) {
         let allListItems = listViewController.sections.flatMap(\.items)
-        let allComponents = componentSections.map(\.components).flatMap { $0 }
+        let allComponents = paymentMethodSections.map(\.paymentMethods).flatMap { $0 }
```
Comment (descorp) [inline AdyenDropIn/Components/PaymentMethodListComponent.swift:104]: `allComponents` -> `allPaymentMethods` ?

File: AdyenDropIn/Models/DropInComponentStyle.swift [hunk 3]
```diff
@@ -25,14 +25,8 @@ public extension DropInComponent {
...
-        /// Indicates any `RedirectComponent` UI style.
-        public var redirectComponent: RedirectComponentStyle?
-        
-        /// Indicates the await action component UI style.
-        public var awaitComponent = AwaitComponentStyle()
-
-        /// Indicates the voucher action component UI style.
-        public var voucherComponent = VoucherComponentStyle()
+        /// Indicates the UI configuration of Action Components
+        public var actionComponents = AdyenActionComponent.Style()
```
Comment (descorp) [inline AdyenDropIn/Models/DropInComponentStyle.swift:29]: This is nice :)

File: AdyenActions/AdyenActionComponent.swift [hunk 4]
```diff
@@ -130,42 +118,82 @@ public final class AdyenActionComponent: ActionComponent, Localizable {
...
         component.presentationDelegate = presentationDelegate
         component.localizationParameters = localizationParameters
         
         component.handle(action)
-        qrCodeComponent = component
+        currentActionComponent = component
     }
 }
+
+public extension AdyenActionComponent {
```
Comment (descorp) [inline AdyenActions/AdyenActionComponent.swift:164]: This extension could be in it's own file

File: AdyenDropIn/Components/PaymentMethodListComponent.swift (hunk 1 above)
Comment (kaphacius) [reply AdyenDropIn/Components/PaymentMethodListComponent.swift:61]: I'm not sure if we need that. To me it seems unnecessary, since the disclosure indicator usually means using native navigation, and something sliding from the right. Maybe removing it is okay

File: AdyenDropIn/Components/PaymentMethodListComponent.swift (hunk 2 above)
Comment (kaphacius) [reply AdyenDropIn/Components/PaymentMethodListComponent.swift:104]: good find

File: AdyenActions/AdyenActionComponent.swift (hunk 4 above)
Comment (kaphacius) [reply AdyenActions/AdyenActionComponent.swift:164]: yeah i was just in a hurry :)

File: AdyenActions/AdyenActionComponent.swift [hunk 5]
```diff
@@ -25,24 +25,17 @@ public final class AdyenActionComponent: ActionComponent, Localizable {
...
+    /// :nodoc:
+    public let style: Style
 
     /// :nodoc:
     public var localizationParameters: LocalizationParameters?
     
     /// :nodoc:
-    public init(apiContext: APIContext) {
+    public init(apiContext: APIContext,
+                style: Style?) {
```
Comment (descorp) [inline AdyenActions/AdyenActionComponent.swift:36]: How about `style: Style? = nil` ?

File: AdyenDropIn/Utilities/ComponentManager.swift [hunk 6]
```diff
@@ -56,36 +55,34 @@ internal final class ComponentManager {
...
+        let paidSection = PaymentMethodsSection(header: nil,
+                                                paymentMethods: paymentMethods.paid,
+                                                footer: paidFooter)
 
         // Stored section
-        let storedSection = ComponentsSection(components: storedComponents)
+        // Filter out payment methods without the Ecommerce shopper interaction.
+        let storedSection = PaymentMethodsSection(header: nil,
+                                                  paymentMethods: paymentMethods.stored.filter {
+                                                    $0.supportedShopperInteractions.contains(.shopperPresent)
```
[the remark this answers is not in the document]
Comment (kaphacius) [reply AdyenDropIn/Utilities/ComponentManager.swift:66]: 👍

File: Adyen/Core/Payment Methods/PaymentMethod.swift [hunk 7]
```diff
@@ -52,4 +52,14 @@ public protocol StoredPaymentMethod: PaymentMethod {
     /// The supported types of shopper interaction.
     var supportedShopperInteractions: [ShopperInteraction] { get }
     
+    /// Indicates whether the shopper is present.
+    var isShopperPresent: Bool { get }
```
Comment (kaphacius) [inline Adyen/Core/Payment Methods/PaymentMethod.swift:56]: this has to be in the protocol. if i only add it to the protocol extension - we get a segmentation fault from swift compiler 🤷

File: Adyen/UI/List/ListItem.swift [hunk 8]
```diff
@@ -28,9 +28,6 @@ public class ListItem: FormItem {
     /// The trailing text of the item.
     public var trailingText: String?
     
-    /// A boolean value indicating whether a disclosure indicator should be shown in the item's cell.
-    public var showsDisclosureIndicator: Bool
```
Comment (mohammedDehairy) [inline Adyen/UI/List/ListItem.swift:32]: why is this removed, its needed for the gift card elements that are not clickable, so the disclosure needs to be .none in this case

File: AdyenDropIn/Components/PaymentMethodListComponent.swift (hunk 1 above)
Comment (descorp) [reply AdyenDropIn/Components/PaymentMethodListComponent.swift]: 👍

File: AdyenActions/AdyenActionComponent.swift [hunk 9]
```diff
@@ -114,10 +102,10 @@ public final class AdyenActionComponent: ActionComponent, Localizable {
     }
     
     private func handle(_ action: ThreeDS2ChallengeAction) {
-        guard let threeDS2Component = threeDS2Component else { return }
+        guard let threeDS2Component = currentActionComponent as? ThreeDS2Component else { return }
```
Comment (mohammedDehairy) [inline AdyenActions/AdyenActionComponent.swift:105]: we should make an assertionFailure here in case this condition fails

File: AdyenDropIn/Components/PaymentMethodListComponent.swift [hunk 10]
```diff
@@ -97,13 +95,14 @@ internal final class PaymentMethodListComponent: ComponentLoader, PresentableCom
     ///
     /// - Parameter component: The component for which to start a loading animation.
     internal func startLoading(for component: PaymentComponent) {
-        let allListItems = listViewController.sections.flatMap(\.items)
-        let allComponents = componentSections.map(\.components).flatMap { $0 }
+        let allPaymentMethods = paymentMethodSections.map(\.paymentMethods).flatMap { $0 }
         
-        guard let index = allComponents.firstIndex(where: { $0 === component }) else {
+        guard let index = allPaymentMethods.firstIndex(where: { $0.type == component.paymentMethod.type }) else {
```
Comment (mohammedDehairy) [inline AdyenDropIn/Components/PaymentMethodListComponent.swift:100]: maybe there are multiple payment methods with the same type, like for example gift card PM's I think have the same type `giftcard`, but not sure though, can you check that, or try to find something else to identify the payment method, like adding a UUID to the payment method, generated in the initializer?

File: AdyenDropIn/DropInComponent.swift [hunk 11]
```diff
@@ -257,7 +256,8 @@ public final class DropInComponent: NSObject, PresentableComponent {
 extension DropInComponent: PaymentMethodListComponentDelegate {
     
     /// :nodoc:
-    internal func didSelect(_ component: PaymentComponent, in paymentMethodListComponent: PaymentMethodListComponent) {
+    internal func didSelect(_ paymentMethod: PaymentMethod, in paymentMethodListComponent: PaymentMethodListComponent) {
+        guard let component = componentManager.component(for: paymentMethod) else { return }
```
Comment (mohammedDehairy) [inline AdyenDropIn/DropInComponent.swift:260]: assertion failure in case the condition fails?

File: Adyen/UI/List/ListItem.swift (hunk 8 above)
Comment (descorp) [reply Adyen/UI/List/ListItem.swift:32]: This is turns `chevron` on and off..

Comment (mohammedDehairy) [reply Adyen/UI/List/ListItem.swift:32]: yes, its needed to hide the chevron for gift card items

Comment (descorp) [reply Adyen/UI/List/ListItem.swift:32]: but if we remove it - there will be no chevron on Drop-in, and it is fine [as we discussed above](https://github.com/Adyen/adyen-ios/pull/535#discussion_r649767666)

Comment (mohammedDehairy) [reply Adyen/UI/List/ListItem.swift:32]: but its in the design, right? if the designers are okay with it, then its fine.

Comment (kaphacius) [reply Adyen/UI/List/ListItem.swift:32]: i haven't talked to design yet. i will

File: AdyenActions/AdyenActionComponent.swift (hunk 9 above)
Comment (kaphacius) [reply AdyenActions/AdyenActionComponent.swift:105]: how does `Trying to handle a 3DS2Challenge, but current component is not a ThreeDS2Component` sound?

Comment (mohammedDehairy) [reply AdyenActions/AdyenActionComponent.swift:105]: sounds good to me

File: AdyenDropIn/Components/PaymentMethodListComponent.swift (hunk 10 above)
Comment (kaphacius) [reply AdyenDropIn/Components/PaymentMethodListComponent.swift:100]: how about adding a function to compare payment methods?
```swift
    func isEqual(to other: PaymentMethod) -> Bool {
        switch (self as? StoredPaymentMethod, other as? StoredPaymentMethod) {
        case (.some(let lhs), .some(let rhs)):
            return lhs.identifier == rhs.identifier
        case (.none, .none):
            return self.name == other.name && self.type == other.type
        case (_, _):
            return false
        }
    }
```

**Comment by kaphacius** [comment]
This is parked for now.
